# Estatística com Python
<p style="text-align:justify">
A estatística é uma área da matemática frequentemente vista como complexa e até controversa, mas é essencial no mundo moderno. Afinal, quem hoje em dia não ouviu falar do ChatGPT ou de inteligência artificial? Esses novos modelos de linguagem computacional são construídos com base em modelos estatísticos amplamente estudados e testados. Além disso, empresas utilizam a estatística para tomar decisões, como em pesquisas de mercado ou análises de vendas de determinados produtos. Outros exemplo são as pesquisas eleitorais, realizadas antes de cada eleição, e o Censo, conduzido nos países para entender o comportamento de sua população. Todos esses exemplos destacam a importância da estatística no cenário atual.
</p>

## Então, por onde começar?

<p style="text-align:justify">
A primeira coisa a se entender na estatística é que existe uma ordem a se seguida para uma estudo estatístico, o primeiro passo é fazer o planejamento do estudo, isso requer escolher as perguntas certas a serem feitas ao publico alvo do estudo. Como o objetivo aqui é entender os conceitos das análises estatística, vamos usar um estudo estatístico já pronto, no livro o de "Practical Data Analysis: Case Studies in Business Statistics" de Bryant, P. G. e Smith, M (1995), eles apresentam conjunto de dados muito conhecido pelos cientistas de dados, um estudo feito por um garçom muito dedicado, no qual passou a anotar todas as gorjetas recebida por ele durante um período, sendo assim iremos trabalhar com o conjunto estatístico, "tips" ou "gorjetas", em português.
</p>

<p style="text-align:justify">
Os dados podem ser encontrados no site do <a href="https://www.kaggle.com/datasets/jsphyg/tipping">Kaggle - A Waiter's Tips</a>, aqui podemos encontra o DataFrame em que iremos trabalhar. DataFrame é uma formula tabular de representação dos dados coletados, veremos mais a seguir como é uma DataFrame.
</p>

<p style="text-align:justify">
Além disso vale lembrar que iremos usar a linguagem de programação, Python. Esta incrível linguagem, através de suas enumeras bibliotecas, nos ajudam não somente em uma análise estatística como em muitas tarefas de nosso dia a dia. Porém para este estudo nos usaremos biblioteca duas bibliotecas essenciais para quem eq estuda estatística, a biblioteca <a href="https://numpy.org/">Numpy</a>, que vai nos ajudar com todas as contas estatísticas, e a biblioteca <a href="https://pandas.pydata.org/">Pandas</a> que vai nos ajudara com as contas estatísticas mas também com as visualização dos dados que foram baixados.  pode nos ajudar em muitas tarefas de nosso dia a dia, com automatização de serviços, ou para nosso caso em uma analise estatística. Você precisa já ter instalado o Python em seu PC pois aqui iremos usar uma de suas enumeras bibliotecas o .
</p>

## Visualizando os dados

<p style="text-align:justify">
Logo após baixar os dados é possível ver que estes dados estão em um arquivo <code>.zip</code> e quando extraído vemos um arquivo com uma nova extensão <code>.csv</code>. 
</p>
<br>

![Pasta com os dados](https://github.com/AndersonAyache/EstatisticaComPython/blob/main/image/pasta_dados.png)


<br>

<p style="text-align:justify">
Esta extensão que é chamade de <i>Comma-separated values</i>, que significa que seus valores são armazenados de tal forma que eles ficam separados por virgula. Como mencionado anteriormente vamos usar a biblioteca do Python chamada <code>Pandas</code>.
</p>

<p style="text-align:justify">
Para instalar esta biblioteca basta digitar no seu ambiente virtual:
</p>

>```python
>pip install pandas
>```

<p style="text-align:justify">
    Agora com a biblioteca instalada podemos finalmente dá nossa primeira olhada nos dados, mas para isso é primeiro preciso importar esta biblioteca, e para isso basta digitar o seguinte comando: 
</p>

In [2]:
import pandas as pd

<p style="text-align:justify">
    Pronto agora já temos a biblioteca instalada, sendo assim podemos atribuir a uma variável os dados que baixamos com a extensão <code>.csv</code>
</p>

In [3]:
gorjetas = pd.read_csv('dados/tips.csv')
gorjetas

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

<p style="text-align: justify">
    Neste primeiro contato com os dados é possível ver como os dados estão dispostos em um tabela, por isso a disposição tabular, ou seja, o DataFrame.
</p>
<p style="text-align: justify">
    Podemos ver que cada linha representa um registro feito pela garçom, quanto que cada coluna representa uma informações sobre estes registos. Estas colunas são chamadas de variáveis estatísticas e pode ser classificada de duas maneiras diferentes, como podemos ver a seguir.
</p>

### Variáveis estatísticas

<p style="text-align:justify">
As variáveis estatísticas são classificadas pela natureza que ela representa, sendo assim as variáveis pode ser <b>Quantitativas</b> ou <b>Qualitativa</b>.
</p>

#### Variáveis Quantitativas

<p style="text-align:justify">
    As variáveis quantitativas são todas as variáveis que a gente pode mensurar, ou seja, que se pode contar, porém existem dois tipos de variáveis quantitativas:
    <ul style="list-style-type:disc">
        <li><b>Discretas</b>: são variáveis que assumem apenas valores inteiros, que são limitadas ou finitas, por exemplo:
            <ul style="list-style-type:circle">
                <li>Quantidade de pessoas em uma sala</li>
                <li>Quantidade de filhos</li>
                <li>Quantidade de acertos em uma prova</li>
                <li>Número de votos</li>
            </ul>
        </li>
        <li><b>Continuas</b>: são variáveis que pode assumir um número infinito de valores dentro de um intervalo, por exemplo:
            <ul style="list-style-type:circle">
                <li>A altura de uma pessoa</li>
                <li>O peso de uma pessoa</li>
                <li>A temperatura de um ambiente</li>
                <li>O volume de um recipiente</li>
            </ul>
        </li>
</p>

#### Variáveis Qualitativas

<p style="text-align:justify">
    As variáveis qualitativas são variáveis que trazem uma qualidade do indivíduo, sendo assim são as variáveis que não conseguimos contar. E também possuem dois tipos:
    <ul style="list-style-type:disc">
        <li><b>Ordinal</b>: quando á variável pode ser disposta usando uma ordem, por exemplo:
            <ul style="list-style-type:circle">
                <li>Classe social</li>
                <li>Grau de instrução</li>
                <li>Conceitos como ótimo, bom, regular ou ruim</li>
            </ul>
        </li>
        <li><b>Nominais</b>: quando se não pode colocar uma ordem na variável, por exemplo:
            <ul style="list-style-type:circle">
                <li>Gênero de uma pessoa </li>
                <li>Cor dos olhos</li>
                <li>Local de nascimento</li>
            </ul>
        </li>
</p>

#### Analisando as variáveis dos nossos dados

Com este comando podemos ver todas as colunas/variáveis do nosso DataFrame

In [4]:
gorjetas.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

Vamos começar observando a primeira coluna a coluna `total_bill` e em seguida analisar uma por uma das colunas/variáveis

In [5]:
gorjetas['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

Esta variável trás o total da conta gasto pode cada cliente, e como podemos ver ela se trata de uma variável **quantitativa e contínua**.

In [6]:
gorjetas['tip']

0      1.01
1      1.66
2      3.50
3      3.31
4      3.61
       ... 
239    5.92
240    2.00
241    2.00
242    1.75
243    3.00
Name: tip, Length: 244, dtype: float64

Está variável tem extrema importância para um estudo de previsão, então podemos dizer que esta variável é a **variável alvo** do estudo, pois ela fornece o valor das gorjetas recebida pelo garçom. Ela também é uma variável **quantitativa e contínua**.

In [7]:
gorjetas['sex']

0      Female
1        Male
2        Male
3        Male
4      Female
        ...  
239      Male
240    Female
241      Male
242      Male
243    Female
Name: sex, Length: 244, dtype: object

Já nesta variável podemos ver que ela nao é mensurável, pois ela se trata do gênero da pessoa que consumiu no estabelecimento onde o garçom trabalha, e como o gênero não possui uma hierarquia ou uma ordem esta variável é **qualitativa nominal**

In [8]:
gorjetas['smoker']

0       No
1       No
2       No
3       No
4       No
      ... 
239     No
240    Yes
241    Yes
242     No
243     No
Name: smoker, Length: 244, dtype: object

A variável `smoker` é referente se havia fumantes sentando a mesa, está é outra variável que é **qualitativa nominal**, pois a resposta para ela é sim ou não sem haver uma ordem.

In [9]:
gorjetas['day']

0       Sun
1       Sun
2       Sun
3       Sun
4       Sun
       ... 
239     Sat
240     Sat
241     Sat
242     Sat
243    Thur
Name: day, Length: 244, dtype: object

Na variável `day`, temos os dias da semana, e neste caso existe uma ordem para isso, ou seja, segunda vem antes da terça, que vem antes da quarta e assim sucessivamente, sendo assim está variável e **qualitativa ordinal**.

In [10]:
gorjetas['time']

0      Dinner
1      Dinner
2      Dinner
3      Dinner
4      Dinner
        ...  
239    Dinner
240    Dinner
241    Dinner
242    Dinner
243    Dinner
Name: time, Length: 244, dtype: object

Na variável `time` se tem o tipo de refeição que foi feita, esta variável também não possui uma hierarquia, afinal nenhuma refeição é mais importante que a outra, todas são importantes para se manter bem alimentando, sendo assim, está variável é **qualitativa nominal**.

In [11]:
gorjetas['size']

0      2
1      3
2      3
3      2
4      4
      ..
239    3
240    2
241    2
242    2
243    2
Name: size, Length: 244, dtype: int64

A variável `size` representa a quantidade de pessoas que estavam presentes em cada mesa que o garçom atendeu, como este valor é finito e possível contar está variável e **qualitativa discreta**.

### Frequências

<p style="text-align:justify">
    Agora que sabemos quais sãos as variáveis e os seus tipos é preciso entender o que cada variável trás de informação e também o que a frequência que estas informações parecem. Para isso vamos verificar os valores únicos para cada variável usando uma função chamada <code> unique()</code> e depois vamos contar quantas vezes estes valores únicos aparecem usando a função <code>value_counts()</code>.
</p>

In [12]:
gorjetas['total_bill'].unique()

array([16.99, 10.34, 21.01, 23.68, 24.59, 25.29,  8.77, 26.88, 15.04,
       14.78, 10.27, 35.26, 15.42, 18.43, 14.83, 21.58, 10.33, 16.29,
       16.97, 20.65, 17.92, 20.29, 15.77, 39.42, 19.82, 17.81, 13.37,
       12.69, 21.7 , 19.65,  9.55, 18.35, 15.06, 20.69, 17.78, 24.06,
       16.31, 16.93, 18.69, 31.27, 16.04, 17.46, 13.94,  9.68, 30.4 ,
       18.29, 22.23, 32.4 , 28.55, 18.04, 12.54, 10.29, 34.81,  9.94,
       25.56, 19.49, 38.01, 26.41, 11.24, 48.27, 13.81, 11.02, 17.59,
       20.08, 16.45,  3.07, 20.23, 15.01, 12.02, 17.07, 26.86, 25.28,
       14.73, 10.51, 27.2 , 22.76, 17.29, 19.44, 16.66, 10.07, 32.68,
       15.98, 34.83, 13.03, 18.28, 24.71, 21.16, 28.97, 22.49,  5.75,
       16.32, 22.75, 40.17, 27.28, 12.03, 12.46, 11.35, 15.38, 44.3 ,
       22.42, 20.92, 15.36, 20.49, 25.21, 18.24, 14.31, 14.  ,  7.25,
       38.07, 23.95, 25.71, 17.31, 29.93, 10.65, 12.43, 24.08, 11.69,
       13.42, 14.26, 15.95, 12.48, 29.8 ,  8.52, 14.52, 11.38, 22.82,
       19.08, 20.27,

É possível ver que existe muitos valores únicos anotados pelo garçom, vamos contar quantos valores únicos existem usando outra função chamada `len()`.

In [13]:
len(gorjetas['total_bill'].unique())

229

São 229 valores únicos inseridos pelo garçom, mas é possível quer que alguns se repente, afinal ele anoutou 244 gorjetas recebidas.

Para ver estes valores que se repente vamos usar a função mencionada anteriormente a `value_counts()`

In [14]:
gorjetas['total_bill'].value_counts()


total_bill
13.42    3
21.01    2
20.69    2
10.33    2
10.34    2
        ..
29.03    1
27.18    1
22.67    1
17.82    1
18.78    1
Name: count, Length: 229, dtype: int64

Agora vamos fazer isso para cada variável, começando pela variável `tip`.

In [15]:
gorjetas['tip'].unique()

array([ 1.01,  1.66,  3.5 ,  3.31,  3.61,  4.71,  2.  ,  3.12,  1.96,
        3.23,  1.71,  5.  ,  1.57,  3.  ,  3.02,  3.92,  1.67,  3.71,
        3.35,  4.08,  2.75,  2.23,  7.58,  3.18,  2.34,  4.3 ,  1.45,
        2.5 ,  2.45,  3.27,  3.6 ,  3.07,  2.31,  2.24,  2.54,  3.06,
        1.32,  5.6 ,  6.  ,  2.05,  2.6 ,  5.2 ,  1.56,  4.34,  3.51,
        1.5 ,  1.76,  6.73,  3.21,  1.98,  3.76,  2.64,  3.15,  2.47,
        1.  ,  2.01,  2.09,  1.97,  3.14,  2.2 ,  1.25,  3.08,  4.  ,
        2.71,  3.4 ,  1.83,  2.03,  5.17,  5.85,  3.25,  4.73,  3.48,
        1.64,  4.06,  4.29,  2.55,  5.07,  1.8 ,  2.92,  1.68,  2.52,
        4.2 ,  1.48,  2.18,  2.83,  6.7 ,  2.3 ,  1.36,  1.63,  1.73,
        2.74,  5.14,  3.75,  2.61,  4.5 ,  1.61, 10.  ,  3.16,  5.15,
        3.11,  3.55,  3.68,  5.65,  6.5 ,  4.19,  2.56,  2.02,  1.44,
        3.41,  5.16,  9.  ,  1.1 ,  3.09,  1.92,  1.58,  2.72,  2.88,
        3.39,  1.47,  1.17,  4.67,  5.92,  1.75])

In [16]:
len(gorjetas['tip'].unique())

123

In [17]:
gorjetas['tip'].value_counts()


tip
2.00    33
3.00    23
4.00    12
2.50    10
5.00    10
        ..
1.47     1
1.17     1
4.67     1
5.92     1
1.75     1
Name: count, Length: 123, dtype: int64

Agora para a variável sexo/gênero.

In [18]:
gorjetas['sex'].unique()

array(['Female', 'Male'], dtype=object)

In [19]:
len(gorjetas['sex'].unique())

2

In [20]:
gorjetas['sex'].value_counts()

sex
Male      157
Female     87
Name: count, dtype: int64

Para a variável fumantes a mesa

In [21]:
gorjetas['smoker'].unique()

array(['No', 'Yes'], dtype=object)

In [22]:
len(gorjetas['smoker'].unique())

2

In [23]:
gorjetas['smoker'].value_counts()

smoker
No     151
Yes     93
Name: count, dtype: int64

Para variável dia da semana

In [24]:
gorjetas['day'].unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [25]:
len(gorjetas['day'].unique())

4

In [26]:
gorjetas['day'].value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

Para a hora da refeição

In [27]:
gorjetas['time'].unique()

array(['Dinner', 'Lunch'], dtype=object)

In [29]:
len(gorjetas['time'].unique())

2

In [28]:
gorjetas['time'].value_counts()

time
Dinner    176
Lunch      68
Name: count, dtype: int64

E por fim a quantidade de pessoas sentada a mesa

In [30]:
gorjetas['size'].unique()

array([2, 3, 4, 1, 6, 5])

In [31]:
len(gorjetas['size'].unique())

6

In [32]:
gorjetas['size'].value_counts()

size
2    156
3     38
4     37
5      5
1      4
6      4
Name: count, dtype: int64